File: site-glue/js/src/main/scala/gramaire/site/EngineClient.scala
package gramaire.site

import org.scalajs.dom
import scala.scalajs.js
import scala.scalajs.js.annotation.JSExportTopLevel
import scala.collection.mutable

// Main-thread client for the engine Web Worker (`EngineWorkerMain`). Runs the
// parse off-thread and hands back a Promise, so the Lab's live re-parse never
// janks on a big grammar. The caller (the Lab) already does latest-wins by
// sequence number, so this stays a thin request/response bridge; a
// superseded request simply has its result ignored upstream.
// Ported from site/src/lib/engine-client.ts.
object EngineClient:

  private var worker: dom.Worker = null
  private var nextId: Int = 1
  private val pending =
    mutable.Map.empty[Int, (GramaireRuntime.GramaireParseResult => Unit, Any => Unit)]

  private def ensureWorker(): dom.Worker =
    if worker != null then worker
    else
      val url = new dom.URL("./engine-worker.mjs", js.`import`.meta.url.asInstanceOf[String])
      val opts = js.Dynamic.literal(`type` = "module").asInstanceOf[dom.WorkerOptions]
      val w = new dom.Worker(url.toString, opts)
      w.onmessage = (e: dom.MessageEvent) =>
        val data = e.data.asInstanceOf[js.Dynamic]
        val id = data.id.asInstanceOf[Int]
        pending.remove(id).foreach { case (resolve, reject) =>
          if !js.isUndefined(data.error) then reject(new Exception(data.error.asInstanceOf[String]))
          else resolve(data.result.asInstanceOf[GramaireRuntime.GramaireParseResult])
        }
      worker = w
      w

  /** Parse off the main thread. Rejects if the worker errors; the Lab's latest-wins guard discards
    * results from superseded edits.
    */
  @JSExportTopLevel("parseInWorker")
  def parseInWorker(
      source: String,
      input: String,
      method: js.UndefOr[GramaireRuntime.GramaireMethod] = js.undefined
  ): js.Promise[GramaireRuntime.GramaireParseResult] =
    val w = ensureWorker()
    val id = nextId
    nextId += 1
    js.Promise[GramaireRuntime.GramaireParseResult]((resolve, reject) =>
      pending.update(id, (r => resolve(r), e => reject(e)))
      w.postMessage(
        js.Dynamic
          .literal(id = id, source = source, input = input, method = method.getOrElse("Canonical"))
      )
    )

File: site/src/lib/engine-client.ts
// Main-thread client for the engine Web Worker (engine-worker.ts). Runs the
// parse off-thread and hands back a Promise, so the Lab's live re-parse never
// janks on a big grammar. The caller (the Lab) already does latest-wins by
// sequence number, so this stays a thin request/response bridge; a superseded
// request simply has its result ignored upstream.
import type { GramaireMethod, GramaireParseResult } from "./gramaire-runtime.ts";

type Reply =
  { id: number; result: GramaireParseResult } | { id: number; error: string };

let worker: Worker | null = null;
let nextId = 1;
const pending = new Map<
  number,
  { resolve: (r: GramaireParseResult) => void; reject: (e: unknown) => void }
>();

function ensureWorker(): Worker {
  if (worker) return worker;
  worker = new Worker(new URL("./engine-worker.ts", import.meta.url), {
    type: "module",
  });
  worker.onmessage = (e: MessageEvent<Reply>) => {
    const p = pending.get(e.data.id);
    if (!p) return;
    pending.delete(e.data.id);
    if ("error" in e.data) p.reject(new Error(e.data.error));
    else p.resolve(e.data.result);
  };
  return worker;
}

/** Parse off the main thread. Rejects if the worker errors; the Lab's
 * latest-wins guard discards results from superseded edits. */
export function parseInWorker(
  source: string,
  input: string,
  method: GramaireMethod = "Canonical",
): Promise<GramaireParseResult> {
  const w = ensureWorker();
  const id = nextId++;
  return new Promise((resolve, reject) => {
    pending.set(id, { resolve, reject });
    w.postMessage({ id, source, input, method });
  });
}
